File: leetcode_py/Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverse(self, p1, k):
        cur = p1.next
        for i in range(k - 1):
            next = cur.next
            cur.next = next.next
            next.next = p1.next  # notice not cur!
            p1.next = next
        return cur

    def testkgroup(self, start_pre, k):
        for i in range(k):
            start_pre = start_pre.next
            if not start_pre:
                return False
        return True

    # @param {ListNode} head
    # @param {integer} k
    # @return {ListNode}
    def reverseKGroup(self, head, k):
        if head == None or head.next == None or k <= 1:
            return head
        dummy = ListNode(0)
        dummy.next = head
        p1 = dummy
        while self.testkgroup(p1, k):
            p1 = self.reverse(p1, k)
        return dummy.next


class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        list = head
        if list == None or k <= 1 or list.next == None:
            return list
        dummy = ListNode(0)
        stack = []
        pre = dummy
        p = head
        dummy.next = head

        while p is not None:
            for i in range(k):
                stack.append(p)
                p = p.next
                if p == None:
                    break

            if len(stack) == k:
                for i in range(k):
                    pre.next = stack.pop()
                    pre = pre.next
            else:
                for i in range(len(stack)):
                    pre.next = stack[i]
                    pre = pre.next

        pre.next = None
        return dummy.next


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


if __name__ == '__main__':
    a = Solution()
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    tmp = a.reverseKGroup(head, 2)
    while tmp:
        print tmp.val
        tmp = tmp.next

File: leetcode_js/Reverse Nodes in k-Group.js
//25. Reverse Nodes in k-Group My Submissions Question
//Total Accepted: 51020 Total Submissions: 191876 Difficulty: Hard
//Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.
//
//If the number of nodes is not a multiple of k then left-out nodes in the end should remain as it is.
//
//You may not alter the values in the nodes, only nodes itself may be changed.
//
//Only constant memory is allowed.
//
//For example,
//Given this linked list: 1->2->3->4->5
//
//For k = 2, you should return: 2->1->4->3->5
//
//For k = 3, you should return: 3->2->1->4->5
//

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
function ListNode(val) {
    this.val = val;
    this.next = null;
}
var reverseList = function (res, k) {
    var cur = res.next
    for (var i = 0; i < k - 1; i++) {
        next = cur.next
        cur.next = next.next
        next.next = res.next
        res.next = next
    }
    return cur;
}
var reverseKGroup = function (head, k) {
    var pre = new ListNode(0),
        res = pre;
    pre.next = head;
    while (true) {
        var next = pre;
        for (var i = 0; i < k; i++) {
            if (next.next !== null) next = next.next;
            else return res.next;
        }
        pre = reverseList(pre, k);
    }
}
var tmphead = new ListNode(1);
tmphead.next = new ListNode(2);
document.writeln(reverseKGroup(tmphead, 2));